### PR TITLE
LogEmitter System.Diagnostics.Tracing extension project

### PR DESCRIPTION
**Pull request by CodeBlanch**
## Changes

* Adds extensions for System.Diagnostics.Tracing

## Proposed Public API

```diff
namespace OpenTelemetry.Logs
{
    public class OpenTelemetryLoggerProvider
    {
+       public bool IncludeFormattedMessage { get; }
    }
}
```

## TODOs

* [ ] Appropriate `CHANGELOG.md` updated for non-trivial changes
* [ ] Changes in public API reviewed

### CODE REVIEW

File: src/OpenTelemetry/Logs/LogRecord.cs [hunk 1]
```diff
@@ -36,6 +36,29 @@ public sealed class LogRecord
 
         private List<object?>? bufferedScopes;
 
+        public LogRecord(
```
Comment (reyang) [inline src/OpenTelemetry/Logs/LogRecord.cs:39]: How about having a ctor which takes no parameter, and have public setters? I think this makes it easier to add additional fields in the future without having to introduce new ctors or struggle with the ordering of parameters.

File: src/OpenTelemetry/Logs/OpenTelemetryLoggerProvider.cs [hunk 2]
```diff
@@ -112,6 +112,12 @@ public ILogger CreateLogger(string categoryName)
             return logger;
         }
 
+        /// <summary>
+        /// Create a <see cref="LogEmitter"/>.
+        /// </summary>
+        /// <returns><see cref="LogEmitter"/>.</returns>
+        public LogEmitter CreateEmitter() => new(this);
```
Comment (reyang) [inline src/OpenTelemetry/Logs/OpenTelemetryLoggerProvider.cs:119]: Do you consider LogEmitter as a pub-sub model? (similar like ActivitySource - which can be subscribed by multiple listeners)

**Comment by CodeBlanch** [comment]
This now also includes `LogRecordPool`. Blame @reyang for the scope creep 🤣

File: src/OpenTelemetry/Batch.cs [hunk 3]
```diff
@@ -50,9 +53,13 @@ public Batch(T[] items, int count)
             this.Count = this.targetCount = count;
         }
 
-        internal Batch(T item)
+        /// <summary>
+        /// Initializes a new instance of the <see cref="Batch{T}"/> struct.
+        /// </summary>
+        /// <param name="item">The item to store in the batch.</param>
+        public Batch(T item)
```
Comment (CodeBlanch) [inline src/OpenTelemetry/Batch.cs:60]: I decided to make this public. So that [ReentrantExportProcessor<T>](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/blob/827eb256890889aeb72d0d641beb92ebd9819b94/src/OpenTelemetry.Exporter.Geneva/ReentrantExportProcessor.cs#L28) can be updated to participate in pooling without needing reflection.

File: src/OpenTelemetry/Logs/LogRecordPool.cs [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+using System;
+using System.Threading;
+using OpenTelemetry.Internal;
+
+namespace OpenTelemetry.Logs
+{
+    /// <summary>
+    /// Manages a pool of <see cref="LogRecord"/> instances.
+    /// </summary>
+    public sealed class LogRecordPool
```
Comment (CodeBlanch) [inline src/OpenTelemetry/Logs/LogRecordPool.cs:28]: I didn't spend a whole lot of time on this algorithm. Just wanted to get something workable in place that we can improve as we go. There is a fast-path using a `[ThreadStatic]` for simple processors executing on the same thread. For the batch case there is a shared pool which is used after the `[ThreadStatic]`.

File: src/OpenTelemetry/Batch.cs (hunk 3 above)
Comment (reyang) [reply src/OpenTelemetry/Batch.cs:60]: FYI - there is a related spec PR https://github.com/open-telemetry/opentelemetry-specification/pull/2452 which I'm trying to help out.

File: src/OpenTelemetry/BatchExportProcessor.cs [hunk 5]
```diff
@@ -268,7 +277,10 @@ private void ExporterProc()
 
                 if (this.circularBuffer.Count > 0)
                 {
-                    using (var batch = new Batch<T>(this.circularBuffer, this.maxExportBatchSize))
+                    using (var batch = new Batch<T>(this.circularBuffer, this.maxExportBatchSize)
+                    {
+                        CleanupAction = cleanupAction,
```
Comment (reyang) [inline src/OpenTelemetry/BatchExportProcessor.cs:282]: What if there are multiple batch processors? Or another processor (e.g. simple) after a batch one?

Comment (CodeBlanch) [reply src/OpenTelemetry/BatchExportProcessor.cs:292]: Tricky tricky! I added reference tracking to solve it. Check it out, LMK your thoughts.

File: examples/AspNetCore/Controllers/WeatherForecastController.cs [hunk 6]
```diff
@@ -54,11 +60,23 @@ public IEnumerable<WeatherForecast> Get()
...
+        // Log using LogEmitter API.
+        var logRecord = LogRecordPool.Rent();
+
+        logRecord.CategoryName = "WeatherForecasts";
+        logRecord.LogLevel = LogLevel.Information;
+        logRecord.FormattedMessage = "WeatherForecasts generated.";
+        logRecord.StateValues = new List<KeyValuePair<string, object?>>() { new KeyValuePair<string, object?>("count", forecast.Length) };
+        logRecord.SetActivityContext(Activity.Current);
+
+        this.logEmitter.Log(logRecord);
```
Comment (reyang) [inline examples/AspNetCore/Controllers/WeatherForecastController.cs:78]: What would be the semantic / expected behavior for the following code snippet? Or it is considered as a user bug (they need to clearly understand the ownership contract)?

```csharp
var logRecord = LogRecordPool.Rent();
...
logRecord.FormattedMessage = "old";
this.logEmitter.Log(logRecord);
logRecord.FormattedMessage = "new";
this.logEmitter.Log(logRecord);
```

Comment (reyang) [reply examples/AspNetCore/Controllers/WeatherForecastController.cs:78]: I'm bit concerned about the API design here. For example, the timestamp was generated during `Rent()`?

File: src/OpenTelemetry/Logs/LogRecordPool.cs [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+            Guard.ThrowIfOutOfRange(size, min: 1);
+
+            current = new LogRecordPool(size);
+        }
+
+        /// <summary>
+        /// Rent a <see cref="LogRecord"/> from the pool.
+        /// </summary>
+        /// <returns><see cref="LogRecord"/>.</returns>
+        public static LogRecord Rent() => current.RentCore(clearIfReused: true);
```
Comment (reyang) [inline src/OpenTelemetry/Logs/LogRecordPool.cs:62]: I don't have a clear position right now - would we prefer to do the clean up here or earlier (e.g. when the ref count goes to zero)? There might be small perf differences (e.g. I imagine `Rent()` is always called synchronously, while DecRef might be called asynchronously - from the exporter thread).

File: src/OpenTelemetry/Batch.cs (hunk 3 above)
Comment (CodeBlanch) [reply src/OpenTelemetry/Batch.cs]: OK it sounds like we are working to pull `ReentrantExportProcessor<T>` into the main repo so I reverted this change.

File: examples/AspNetCore/Controllers/WeatherForecastController.cs (hunk 6 above)
Comment (CodeBlanch) [reply examples/AspNetCore/Controllers/WeatherForecastController.cs]: Pushed a new iteration. Pooling is now encapsulated inside `OpenTelemetryLogger` and `LogEmitter` so the above is not possible. Also added a way to pass up to 8 tags/attributes without allocation. Check it out, LMK your thoughts on the new version.

File: src/OpenTelemetry/Logs/LogRecordPool.cs (hunk 7 above)
Comment (CodeBlanch) [reply src/OpenTelemetry/Logs/LogRecordPool.cs]: In the new iteration there isn't much to clear, log data is set fully each time (overriding what was there before). There is a bit of cleanup for the new reusable tag/attribute storage which I put on the return to the pool.

**Comment by reyang** [comment]
> @reyang Some stress test runs...
> 
> * Before
> 
> ```shell
> Running (concurrency = 20, prometheusEndpoint = http://localhost:9184/metrics/), press <Esc> to stop...
> Stopping the stress test...
> * Total Loops: 8,040,652,772
> * Average Loops/Second: 36,874,594
> * Average CPU Cycles/Loop: 1,342
> 
> Running (concurrency = 20, prometheusEndpoint = http://localhost:9184/metrics/), press <Esc> to stop...
> Stopping the stress test...
> * Total Loops: 8,022,938,366
> * Average Loops/Second: 35,295,762
> * Average CPU Cycles/Loop: 1,431
> 
> Running (concurrency = 20, prometheusEndpoint = http://localhost:9184/metrics/), press <Esc> to stop...
> Stopping the stress test...
> * Total Loops: 8,026,902,532
> * Average Loops/Second: 33,723,789
> * Average CPU Cycles/Loop: 1,500
> ```
> 
> * After
> 
> ```shell
> Running (concurrency = 20, prometheusEndpoint = http://localhost:9184/metrics/), press <Esc> to stop...
> Stopping the stress test...
> * Total Loops: 8,033,179,136
> * Average Loops/Second: 57,206,597
> * Average CPU Cycles/Loop: 977
> 
> Running (concurrency = 20, prometheusEndpoint = http://localhost:9184/metrics/), press <Esc> to stop...
> Stopping the stress test...
> * Total Loops: 8,025,575,616
> * Average Loops/Second: 40,543,653
> * Average CPU Cycles/Loop: 1,385
> 
> Running (concurrency = 20, prometheusEndpoint = http://localhost:9184/metrics/), press <Esc> to stop...
> Stopping the stress test...
> * Total Loops: 8,032,760,075
> * Average Loops/Second: 59,843,255
> * Average CPU Cycles/Loop: 939
> ```
> 
> I'm going to mess with detecting if the processor chain needs reference counting, but so far optimistic about the results.

Nice, so it's already paying off (and much more)!!!

File: examples/AspNetCore/Controllers/WeatherForecastController.cs [hunk 8]
```diff
@@ -54,11 +60,25 @@ public IEnumerable<WeatherForecast> Get()
...
         .ToArray();
 
+        // Log using ILogger API.
         this.logger.LogInformation(
             "WeatherForecasts generated {count}: {forecasts}",
             forecast.Length,
             forecast);
 
+        // Log using LogEmitter API.
+        this.logEmitter.Log(
```
Comment (alanwest) [inline examples/AspNetCore/Controllers/WeatherForecastController.cs:70]: I wonder if it makes sense to demonstrate this here? My understanding is that this isn't really meant to be used by end users, but by authors of extensions for various logging libraries like serilog or log4net. Have you had a chance to experiment with the API in the context of some other log framework?

File: src/OpenTelemetry/.publicApi/net462/PublicAPI.Unshipped.txt [hunk 9]
```diff
@@ -1,4 +1,49 @@
 #nullable enable
+OpenTelemetry.Logs.LogEmitter
```
[the remark this answers is not in the document]
Comment (reyang) [reply src/OpenTelemetry/.publicApi/net462/PublicAPI.Unshipped.txt:2]: Here is my suggestion:
1. Use this PR to focus on the LogRecord improvements (including polling, ref count, setters) - this would give good perf benefits, and set a cornerstone for LogEmitter or any bridge with Serilog/NLog/EventSource/etc.
2. LogEmitter can either be a separate PR or included here, as long as the scope of the change is not too big + is self-contained (e.g. if LogEmitter is need to make meaningful perf tests, probably fine to keep it, but it has to be kept `internal`)
3. Before the spec becomes stable, we can either keep LogEmitter `internal` (for our own testing purpose), or making it public AND having it in a separate package (similar to OTLP Log Exporter) - if we know someone who is interested in trying it out.

[the remark this answers is not in the document]
Comment (alanwest) [reply src/OpenTelemetry/.publicApi/net462/PublicAPI.Unshipped.txt:2]: Sounds good. 👍 to making `LogEmitter` internal and leaving it where it is in this PR.

File: src/OpenTelemetry/.publicApi/net462/PublicAPI.Unshipped.txt [hunk 10]
```diff
@@ -1,4 +1,49 @@
...
+OpenTelemetry.Logs.LogEmitter.Log(in OpenTelemetry.Logs.LogRecordData data, in OpenTelemetry.Logs.LogRecordAttributeList attributes = default(OpenTelemetry.Logs.LogRecordAttributeList)) -> void
+OpenTelemetry.Logs.LogEmitter.LogEmitter(OpenTelemetry.Logs.OpenTelemetryLoggerProvider! loggerProvider) -> void
 OpenTelemetry.Logs.LogRecord.FormattedMessage.set -> void
+OpenTelemetry.Logs.LogRecord.GetDataRef() -> OpenTelemetry.Logs.LogRecordData
 OpenTelemetry.Logs.LogRecord.State.set -> void
 OpenTelemetry.Logs.LogRecord.StateValues.set -> void
+OpenTelemetry.Logs.LogRecordAttributeList
```
Comment (alanwest) [inline src/OpenTelemetry/.publicApi/net462/PublicAPI.Unshipped.txt:9]: Since .NET uses the term `Tag` for metrics and traces, should we stay consistent and call 'em tags for logs as well?

File: src/OpenTelemetry/.publicApi/net462/PublicAPI.Unshipped.txt (hunk 9 above)
[the remark this answers is not in the document]
Comment (CodeBlanch) [reply src/OpenTelemetry/.publicApi/net462/PublicAPI.Unshipped.txt:2]: * Let's make it internal and keep where it is. We can figure out where to move it if needed before it goes public?

* What shall I do with `LogRecordData` + `LogRecordAttributeList`?
  * If `LogRecordData` goes internal, so to will `LogRecord.GetDataRef()` which gives exporters a way to get at the data more quickly and also gives users a way to mutate _everything_ on `LogRecord` (which has been requested).
  * If `LogRecordAttributeList` goes internal, no real downside without `LogEmitter`. I do have this in there...
    https://github.com/open-telemetry/opentelemetry-dotnet/blob/2114dcaae6026b4927f5e38360808436329d1f83/src/OpenTelemetry/Logs/OpenTelemetryLogger.cs#L93

    What that allows you to do is pass `LogRecordAttributeList` as `TState` to `ILogger.Log<TState>` and get boxing/allocation-free tags. If it goes internal, users won't be able to easily build upon that perf-path.

File: src/OpenTelemetry/.publicApi/net462/PublicAPI.Unshipped.txt (hunk 10 above)
Comment (CodeBlanch) [reply src/OpenTelemetry/.publicApi/net462/PublicAPI.Unshipped.txt:9]: I struggled with that one. We have `State` & `StateValues` on `LogRecord` today. The most consistent thing might be to name this `LogRecordStateValueList` and not introduce another concept? I ended up going with "attributes" because that is what the spec calls them and the `LogEmitter` API right now is pretty close to spec but open to whatever.

Comment (alanwest) [reply src/OpenTelemetry/.publicApi/net462/PublicAPI.Unshipped.txt:9]: 🤔 yea I hear ya on the struggle. I think sticking to `Attribute` to align with the spec might make sense in this context. Though also maybe this is a good enough reason to make `LogRecordAttributeList` internal for now. I guess I'm not too clear yet how all this may play out in the context of a logging framework besides ILogger.

Comment (CodeBlanch) [reply src/OpenTelemetry/.publicApi/net462/PublicAPI.Unshipped.txt:9]: Pretty much, just ignore `ILogger` 😄 What `LogEmitter` does is take a `LogRecord` from the pool, applies the `LogRecordData` + `LogRecordAttributeList` to it, and then sends it at the `LogProcessor` at the root of the `OpenTelemetryLoggerProvider`. It doesn't really flow through `ILogger` at all.

I haven't tested this yet, but to help illustrate here is the Serilog sink I scratched out to mess with the `LogEmitter` API:

```csharp
    public static class OpenTelemetrySerilogSinkExtensions
    {
        public static LoggerConfiguration OpenTelemetry(
            this LoggerSinkConfiguration loggerConfiguration,
            OpenTelemetryLoggerProvider openTelemetryLoggerProvider)
        {
            Guard.ThrowIfNull(loggerConfiguration);

            return loggerConfiguration.Sink(new OpenTelemetrySerilogSink(openTelemetryLoggerProvider));
        }
    }

    internal sealed class OpenTelemetrySerilogSink : ILogEventSink
    {
        private readonly LogEmitter logEmitter;

        public OpenTelemetrySerilogSink(OpenTelemetryLoggerProvider openTelemetryLoggerProvider)
        {
            this.logEmitter = new LogEmitter(openTelemetryLoggerProvider);
        }

        public void Emit(LogEvent logEvent)
        {
            Debug.Assert(logEvent != null, "LogEvent was null.");

            LogRecordData data = new(Activity.Current)
            {
                Timestamp = logEvent!.Timestamp.UtcDateTime,
                LogLevel = (LogLevel)(int)logEvent.Level,
                Message = logEvent.MessageTemplate.Text,
                Exception = logEvent.Exception,
            };

            LogRecordAttributeList attributes = default;
            foreach (KeyValuePair<string, LogEventPropertyValue> property in logEvent.Properties)
            {
                // TODO: Serilog supports complex type logging. This is not yet
                // supported in OpenTelemetry.
                if (property.Key == Constants.SourceContextPropertyName)
                {
                    data.CategoryName = property.Value.ToString();
                }
                else if (property.Value is ScalarValue scalarValue)
                {
                    attributes.Add(property.Key, scalarValue.Value);
                }
                else if (property.Value is SequenceValue sequenceValue)
                {
                    IReadOnlyList<LogEventPropertyValue> elements = sequenceValue.Elements;
                    if (elements.Count > 0 && elements[0] is ScalarValue)
                    {
                        object[] values = new object[elements.Count];
                        for (int i = 0; i < elements.Count; i++)
                        {
                            if (elements[i] is ScalarValue value)
                            {
                                values[i] = value.Value;
                            }
                        }

                        attributes.Add(property.Key, values);
                    }
                }
            }

            this.logEmitter.Log(in data, in attributes);
        }
    }
```

Comment (alanwest) [reply src/OpenTelemetry/.publicApi/net462/PublicAPI.Unshipped.txt:9]: Cool, yea this example is great. Is the reason due to size/performance that `LogRecordData` doesn't have an LogRecordAttributeList field? ... Oh, or maybe the reason is that if you were to do `this.logEmitter.Log(in data)` that the `in data` parameter would not preclude the attributes from being modified.

Comment (alanwest) [reply src/OpenTelemetry/.publicApi/net462/PublicAPI.Unshipped.txt:9]: Cool, yea this example is great. Is the reason due to size/performance that `LogRecordData` doesn't have an LogRecordAttributeList field? ... Oh, or maybe the reason is that if you were to do `this.logEmitter.Log(in data)` that the `in data` parameter would not preclude the attributes from being modified.

File: src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/LogRecordExtensions.cs [hunk 11]
```diff
@@ -65,30 +65,31 @@ internal static class LogRecordExtensions
         internal static OtlpLogs.LogRecord ToOtlpLog(this LogRecord logRecord)
         {
             OtlpLogs.LogRecord otlpLogRecord = null;
+            ref LogRecordData data = ref logRecord.GetDataRef();
```
Comment (alanwest) [inline src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/LogRecordExtensions.cs:68]: Just confirming my own understanding of the improvements this PR brings... so this is more performant because accessing properties of the struct results in a `call` vs. `callvirt`, right?

File: src/OpenTelemetry/.publicApi/net462/PublicAPI.Unshipped.txt [hunk 12]
```diff
@@ -1,4 +1,49 @@
...
+OpenTelemetry.Logs.LogRecordData.SpanId.set -> void
+OpenTelemetry.Logs.LogRecordData.Timestamp.get -> System.DateTime
+OpenTelemetry.Logs.LogRecordData.Timestamp.set -> void
+OpenTelemetry.Logs.LogRecordData.TraceFlags.get -> System.Diagnostics.ActivityTraceFlags
+OpenTelemetry.Logs.LogRecordData.TraceFlags.set -> void
+OpenTelemetry.Logs.LogRecordData.TraceId.get -> System.Diagnostics.ActivityTraceId
+OpenTelemetry.Logs.LogRecordData.TraceId.set -> void
+OpenTelemetry.Logs.LogRecordData.TraceState.get -> string?
+OpenTelemetry.Logs.LogRecordData.TraceState.set -> void
+OpenTelemetry.Logs.LogRecordPool
```
Comment (alanwest) [inline src/OpenTelemetry/.publicApi/net462/PublicAPI.Unshipped.txt:46]: Can LogRecordPool be internal for now too if we're keeping LogEmitter and friends internal?

Comment (CodeBlanch) [reply src/OpenTelemetry/.publicApi/net462/PublicAPI.Unshipped.txt:29]: Could be. Only question is do we want to allow users to configure the pool size? The only thing exposed is currently `LogRecordPool.Resize(int size)`.

File: src/OpenTelemetry/.publicApi/net462/PublicAPI.Unshipped.txt (hunk 10 above)
Comment (CodeBlanch) [reply src/OpenTelemetry/.publicApi/net462/PublicAPI.Unshipped.txt]: Originally I had the attributes as a field on `LogRecordData` but I ran into some weird compiler limitations. What I wanted to do was allow people to do this...

```csharp
logEmitter.Log(new()
{
   Message = "Hello world",
   Attributes = { ["count" ] = 1, ["id"] = 18 }
})
```

That would work if "Attributes" was a reference type, but compiler doesn't allow with a value type! I tried a bunch of things, ultimately moved it to a separate parameter. At the moment you can do...

```csharp
logEmitter.Log(
   new() { Message = "Hello world" },
   new() { ["count" ] = 1, ["id"] = 18 }
})
```

Besides that...

* Having it separate keeps the size of `LogRecordData` down. Now `LogRecord`s live in a pool and store a `LogRecordData`. Having the "attributes" as part of `LogRecordData` would mean that pool would take up a bunch more memory.

* I could see people mess it up. Imagine something like this...

   ```csharp
   LogRecord data = new()
   {
      Message = "Hello world"
   };
   
   BuildAttributes(data.Attributes); // This passes a copy. So anything added in the method won't really be added.
   
   logEmitter.Log(in data);
   ```

* Having it separate leaves some room for expansion. What I'm working towards here is potentially using `LogEmitter` to support logging of complex types. Maybe the API will end up like this:

   ```csharp
   LogEmitter.Log(in LogRecordData data, in LogRecordAttributeList attributes = default, in LogRecordObjectWriter objectWriter = default)
   ```

   Basically you always pass in the core sutff, but then you can pass in other things that depend on the case. I have no idea what it will ultimately look like but something I'm thinking about.

At one point I tried a builder pattern like...

```csharp
logEmitter.Log(new()
   .SetMessage("Hello world")
   .AddAttribute("count", 1))
```

But that is tricky to do with a struct. Each time you return the struct it is a new copy so you have to write it like ```public static ref LogRecordBuilder SetMessage(this ref LogRecordBuilder builder)```. Seemed gross 🤣

File: src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/LogRecordExtensions.cs (hunk 11 above)
Comment (CodeBlanch) [reply src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/LogRecordExtensions.cs:68]: This particular change is more to prevent a perf regression! This method used to work something like...

```csharp
ldarg.1 -> call LogRecord.CategoryName.Get() -> ldfld CategoryName_Backing -> ret
```

(Because `LogRecord` is sealed it should already be a `call` instead of `callvirt`.)

Now due to the pooling changes it has to do...

```csharp
ldarg.1 -> call LogRecord.CategoryName.Get() -> ldflda this.logRecordData -> call LogRecordData.CategoryName.Get() -> ldfld CategoryName_Backing -> ret
```

Extra call!

Using `GetDataRef()` it goes back to a single call like...

```csharp
ldflda [address of ref] -> call LogRecordData.CategoryName.Get()-> ldfld CategoryName_Backing -> ret
```

Just an optimization to cut out the extra hops.

True perf really comes down to how the JIT decides to inline things. Here's a [sharplab](https://sharplab.io/#v2:C4LghgzgtgPgAgJgIwFgBQcAMACOSB0ASgK4B2wAllAKb4DCA9lAA4UA21ATgMpcBuFAMbUIAbnTo4AZlwJsdbAG902VbhlwALNgCySABQAZBgHNC1QQ04ATbG1PnLNgJRKVaj3hyCwwaiasATwA5MBpsAF47Bwsra3pffyDQmnE0D08kHBoICDATakjos1ibfB0RPIK0jLV3WrwATn0fPwDOELDqZxra3CRmnKru3tUAX3rsSelcbR0EIxinW3sS5ddldL7OagAzbGM1uIARXzBsazOinf3Vxzj8O9LrU+AwUYzJzO9E9s7wqKXN4JNrJLofb7YIb5QqAs7lSowiF1LYNAYtX5gmg9L5qJr6aEFHp9Kao7BjDyTaYyCDATjEQTAA5LE5nSabEkzLzyTEdFKFRSkkkeADaFWAAAsGNYAJIsNj6cVS2XygDyzEoDFIEHwwQYMtIbAopGNJmcAF1cbUCsBRELhaoINRbeSrep+jgKrkYUp7Q6xc7lXLmAqldLg2x1Zrtbr9YbjaaLW6PDa7cm1E6XRMydnKWSuXJDvcbOy3ca/JxSGA2MyjjZXucnssG8jVG6uVkeaC+V1fenVAHJeH5YrA8OQ1GKFqdXqDUaTaQzZaySSbZEAHzFYsveF0Xn/aitjKZjdb54NkFJHsA7B8avEQ+uld4jSdr3DPvP2qDoMjsMqicNSnGNZ3jBcl37bA1wiTcm1ZYF3yRP1hRPGCz2beFEIKIo7zYB87VzNRszGIA==) if you want to mess with it.

Comment (CodeBlanch) [reply src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/LogRecordExtensions.cs:68]: So what are the real perf gains on this PR?

I will attempt to answer this 😄 

Our log logic was like this:

```csharp
                var record = new LogRecord(
                    provider.IncludeScopes ? this.ScopeProvider : null,
                    DateTime.UtcNow,
                    this.categoryName,
                    logLevel,
                    eventId,
                    provider.IncludeFormattedMessage ? formatter?.Invoke(state, exception) : null,
                    provider.ParseStateValues ? null : state,
                    exception,
                    provider.ParseStateValues ? this.ParseState(state) : null);

                processor.OnEnd(record);

                record.ScopeProvider = null;
```

And is now like this:

```csharp
                var record = LogRecordPool.Rent();

                record.ScopeProvider = provider.IncludeScopes ? this.ScopeProvider : null;
                record.State = provider.ParseStateValues ? null : state;
                record.StateValues = provider.ParseStateValues ? this.ParseState(record, state) : null;

                ref LogRecordData data = ref record.Data;

                data.TimestampBacking = DateTime.UtcNow;
                data.CategoryName = this.categoryName;
                data.LogLevel = logLevel;
                data.EventId = eventId;
                data.Message = provider.IncludeFormattedMessage ? formatter?.Invoke(state, exception) : null;
                data.Exception = exception;

                LogRecordData.SetActivityContext(ref data, Activity.Current);

                processor.OnEnd(record);

                record.ScopeProvider = null;

                // Attempt to return the LogRecord to the pool. This will no-op
                // if a batch exporter has added a reference.
                LogRecordPool.Return(record);
```

Should be slower, right?

Turns out, is only slightly slower (when hitting the `[ThreadStatic]`):

|       Method |     Mean |    Error |   StdDev |  Gen 0 | Allocated |
|------------- |---------:|---------:|---------:|-------:|----------:|
| LogUsingCtor | 38.89 ns | 0.154 ns | 0.137 ns | 0.0102 |     128 B |
| LogUsingPool | 40.37 ns | 0.054 ns | 0.048 ns |      - |         - |

But we eliminated all the memory pressure. I think this is why the stress test is performing better with the pool. Only a few more CPU cycles to use the pool but the GC is asleep freeing up the CPU to process more logs. This is a largely a guess but kind of makes sense?

Now that benchmark is interesting. How is it introducing the pool logic made it only slightly slower than just calling ctor?

Check out this benchmark:

```csharp
    public class CtorBenchmarks
    {
        [Benchmark]
        public TestSmallClass CtorSmallClass()
        {
            return new();
        }

        [Benchmark]
        public TestLargeClass CtorLargeClass()
        {
            return new();
        }

        public class TestSmallClass
        {
            public string? Prop1;
        }

        public class TestLargeClass
        {
            public string? Prop1;
            public string? Prop2;
            public string? Prop3;
            public int Prop4;
            public int Prop5;
            public int Prop6;
            public DateTime Prop7;
            public object? Prop8;
            public object? Prop9;
            public object? Prop10;
            public string? Prop11;
            public string? Prop12;
            public string? Prop13;
        }
    }
```

|         Method |     Mean |     Error |    StdDev |  Gen 0 | Allocated |
|--------------- |---------:|----------:|----------:|-------:|----------:|
| CtorSmallClass | 1.884 ns | 0.0236 ns | 0.0209 ns | 0.0019 |      24 B |
| CtorLargeClass | 3.982 ns | 0.0298 ns | 0.0279 ns | 0.0089 |     112 B |

Number of fields on the class impacts the ctor time. `LogRecord` has a lot of fields so that gives us some wiggle room to execute our pool logic before just calling the ctor would be faster. Interesting, eh?

File: src/OpenTelemetry.Extensions.Serilog/OpenTelemetrySerilogExtensions.cs [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+        /// <summary>
+        /// Adds a sink to Serilog <see cref="LoggerConfiguration"/> which will
+        /// write to OpenTelemetry.
+        /// </summary>
+        /// <param name="loggerConfiguration"><see cref="LoggerSinkConfiguration"/>.</param>
+        /// <param name="openTelemetryLoggerProvider"><see cref="OpenTelemetryLoggerProvider"/>.</param>
+        /// <returns>Supplied <see cref="LoggerConfiguration"/> for chaining calls.</returns>
+        public static LoggerConfiguration OpenTelemetry(
+            this LoggerSinkConfiguration loggerConfiguration,
+            OpenTelemetryLoggerProvider openTelemetryLoggerProvider)
```
Comment (alanwest) [inline src/OpenTelemetry.Extensions.Serilog/OpenTelemetrySerilogExtensions.cs:37]: For some of our other extension methods we've put the extension method in the same namespace as the class that's being extended. So, does it make sense to put this method in the `Serilog.Configuration` namespace, or just the root `Serilog` namespace? Is there an existing precedence for other Serilog extensions?

File: OpenTelemetry.sln [hunk 14]
```diff
@@ -227,7 +227,13 @@ Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "OpenTelemetry.Extensions.Pr
 EndProject
 Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "correlation", "docs\logs\correlation\correlation.csproj", "{9A07D215-90AC-4BAF-BCDB-73D74FD3A5C5}"
 EndProject
-Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "OpenTelemetry.Tests.Stress.Logs", "test\OpenTelemetry.Tests.Stress.Logs\OpenTelemetry.Tests.Stress.Logs.csproj", "{4298057B-24E0-47B3-BB76-C17E81AF6B39}"
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "OpenTelemetry.Tests.Stress.Logs", "test\OpenTelemetry.Tests.Stress.Logs\OpenTelemetry.Tests.Stress.Logs.csproj", "{5FC0660F-3757-4594-806B-4375E06177A3}"
+EndProject
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "OpenTelemetry.Extensions.Serilog", "src\OpenTelemetry.Extensions.Serilog\OpenTelemetry.Extensions.Serilog.csproj", "{F5062ED1-6B59-45FC-8E08-2F5D41A19864}"
+EndProject
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "OpenTelemetry.Extensions.Tracing", "src\OpenTelemetry.Extensions.Tracing\OpenTelemetry.Extensions.Tracing.csproj", "{F1C65913-81EC-4297-A666-A66280E3E1B6}"
```
Comment (alanwest) [inline OpenTelemetry.sln:234]: Tracing is such a commonly used term. I get it comes from System.Diagnostics.Tracing, but is there a better name? 

What about `OpenTelemetry.Extensions.EventSource`? Or if that's too narrow of a name maybe `OpenTelemetry.Extensions.EventTracing`? Or... ugh maybe even `OpenTelemetry.Extensions.SystemDiagnosticsTracing` 🥴?

File: src/OpenTelemetry/Logs/OpenTelemetryLoggerProvider.cs [hunk 15]
```diff
@@ -90,6 +89,11 @@ internal OpenTelemetryLoggerProvider(OpenTelemetryLoggerOptions options)
             }
         }
 
+        /// <summary>
+        /// Gets a value indicating whether or not formatted messages should be included on log messages.
+        /// </summary>
+        public bool IncludeFormattedMessage { get; }
```
Comment (alanwest) [inline src/OpenTelemetry/Logs/OpenTelemetryLoggerProvider.cs:95]: Bugging you again with the "we sure we wanna expose this publicly, yet?" question 😆... I don't have a strong feeling that we shouldn't, just wanna raise it for discussion.

Brainstorming alternatives, would it make sense instead to expose this on the `LogEmitter` class?

At the moment, both the extensions are instantiated with an instance of the logging provider and then
1. store the `IncludeFormattedMessage` setting, then
2. call `CreateEmitter()`

Would it make sense to just instantiate the extensions with an emitter instead of the provider? So like Serilog would look like

```csharp
Log.Logger = new LoggerConfiguration()
    .WriteTo.OpenTelemetry(openTelemetryLoggerProvider.CreateEmitter()) // <- Register OpenTelemetry Serilog sink
    .CreateLogger();
```

Again, I don't have a strong sense that one option is better than the other at this point, just curious of your thoughts.

**Comment by CodeBlanch** [comment]
Closing this PR because everything it was doing has been moved to smaller PRs.
